File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        vector<int> ans(n,0);
        stack<pair<int,int>> st;
        for(int i=n-1;i>=0;i--) {
            if(st.empty()) {
                st.push({temperatures[i],i});
            } else {
                while(st.empty()==false && st.top().first<=temperatures[i]) {st.pop();}
                if(st.empty()==false) {
                    ans[i] = st.top().second;}
                st.push({temperatures[i],i});
            }
        }
        for(int i=n-1;i>=0;i--) {
            if(ans[i]==0) continue;
            ans[i] = ans[i]-i;
        }
        return ans;
    }
};

File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ngl = [0]*n
        ngl_stack = []
        for r in range(n-1,-1,-1):
            while ngl_stack and nums[ngl_stack[-1]]<=nums[r]:
                ngl_stack.pop()
            
            if ngl_stack:
                ngl[r] = ngl_stack[-1]-r
            ngl_stack.append(r)
        return ngl